modules/library/render/src/main/java/org/geotools/renderer/lite/StreamingRenderer.java

```diff
@@ -1067,10 +1095,68 @@ Query getStyleQuery(
             List<ReferencedEnvelope> envelopes = null;
             // enable advanced projection handling with the updated map extent
             if (isAdvancedProjectionHandlingEnabled()) {
+                Map projectionHints = new HashMap();
+                if (isAdvancedProjectionDensificationEnabled()
+                        && !CRS.equalsIgnoreMetadata(featCrs, mapCRS)) {
+                    double tolerance = getAdvancedProjectionDensificationTolerance();
+                    if (tolerance > 0.0) {
+                        ReferencedEnvelope targetEnvelope = envelope;
+                        ReferencedEnvelope sourceEnvelope =
+                                transformEnvelope(targetEnvelope, featCrs);
+                        AffineTransform at = worldToScreenTransform;
+                        AffineTransform screenToWorldTransform = new AffineTransform(at);
+                        screenToWorldTransform.invert();
+                        MathTransform2D crsTransform =
+                                (MathTransform2D)
+                                        CRS.findMathTransform(
+                                                CRS.getHorizontalCRS(featCrs),
+                                                CRS.getHorizontalCRS(mapCRS));
+                        MathTransform2D screenTransform = new AffineTransform2D(at);
+                        MathTransform2D fullTranform =
+                                (MathTransform2D)
+                                        ConcatenatedTransform.create(crsTransform, screenTransform);
+                        MathTransform2D inverseTranform = fullTranform.inverse();
+                        Rectangle2D.Double sourceDomain =
+                                new Rectangle2D.Double(
+                                        sourceEnvelope.getMinX(),
+                                        sourceEnvelope.getMinY(),
+                                        sourceEnvelope.getWidth(),
+                                        sourceEnvelope.getHeight());
+                        WarpBuilder wb = new WarpBuilder(tolerance);
+                        double densifyDistance = 0.0;
+                        int[] actualSplit = wb.getRowColsSplit(fullTranform, sourceDomain);
+                        if (actualSplit == null) {
+                            // alghoritm gave up, we decide to use a fixed distance value
+                            densifyDistance = 5.0;
+                        } else if (actualSplit[0] != 1 || actualSplit[1] != 1) {
+                            Shape shape =
+                                    inverseTranform.createTransformedShape(
+                                            new Rectangle(MAX_PIXELS_DENSIFY, MAX_PIXELS_DENSIFY));
+
+                            Rectangle2D shapeBounds = shape.getBounds2D();
+                            densifyDistance =
```

Comment:
I don't understand what is going on here :-D

Reply:
It is an attempt to "translate" MAX_PIXELS_DENSIFY (5 pixels) to a source CRS distance (e.g 1 degree in EPSG:4326) , but it's not working well, I will try a different approach. The idea is to use that distance when WarpBuilder fails and as a minimum distance.

Reply:
Looks better now, thanks!